### ISSUE TITLE
stickyConnPool.Remove not really do remove any conn?

### ISSUE THREAD

**Issue by dunixd** [go-redis/redis#260]
What is the different between Put and Remove? Is it designed this way? It seem should call remove in Remove.

```
func (p *stickyConnPool) Put(cn *conn) error {
	defer p.mx.Unlock()
	p.mx.Lock()
	if p.closed {
		return errClosed
	}
	if p.cn != cn {
		panic("p.cn != cn")
	}
	return nil
}

func (p *stickyConnPool) remove(reason error) (err error) {
	err = p.pool.Remove(p.cn, reason)
	p.cn = nil
	return err
}

func (p *stickyConnPool) Remove(cn *conn, _ error) error {
	defer p.mx.Unlock()
	p.mx.Lock()
	if p.closed {
		return errClosed
	}
	if p.cn == nil {
		panic("p.cn == nil")
	}
	if cn != nil && p.cn != cn {
		panic("p.cn != cn")
	}
	return nil
}
```

**Comment by vmihailenco** [comment]
`stickyConnPool` is used with PubSub and Transaction, i.e. where connection has some state and requires special attention. For example, PubSub manually closes connection and resubscribes to channels - https://github.com/go-redis/redis/blob/master/pubsub.go#L235-L249

**Comment by dunixd** [comment]
Thank you!
I met a problem when I use `Client.Watch`, redis server close the connection after a idle time, I received an error

> Watch err: EOF
> Unwatch failed: EOF

this bad connection will not be removed by `stickyConnPool` but return to `connPool` as a good one, and reset it's last use time. So next time when I fetch one for `stickyConnPool`, I still maybe get that bad one. And get the error forever

> 2016/02/26 21:24:28 redis: Unwatch failed: write tcp 10.143.160.159:10583->10.117.81.193:6959: write: broken pipe
> 2016/02/26 21:24:36 redis: Unwatch failed: write tcp 10.143.160.159:10583->10.117.81.193:6959: write: broken pipe
> 2016/02/26 21:24:41 redis: Unwatch failed: write tcp 10.143.160.159:10583->10.117.81.193:6959: write: broken pipe
> 2016/02/26 21:24:53 redis: Unwatch failed: write tcp 10.143.160.159:10583->10.117.81.193:6959: write: broken pipe

But the connection is over actually. Maybe the only solution is the bad connection idle before coming out from `connPool`.

So I think maybe stickyConnPool.Remove should remove the bad connection for the Transaction is already failed when Net error occurs.

**Comment by vmihailenco** [comment]
Sorry, I did not have time to look at this yet. I will try to find time tomorrow.

**Comment by vmihailenco** [comment]
@dunixd can you please show the code?

Technically you should not keep `Multi` connection idle for long time. Workflow with `Multi` should be:
- watch for key
- exec some commands
- close multi

See https://godoc.org/gopkg.in/redis.v3#example-Client-Watch for a good example.

**Comment by dunixd** [comment]
The workflow I follow is correct. The connection idle in the `connPool`, before fetched by `stickyConnPool`. It is because my wrong idle time setting. But when redis server close a connection for some reason, maybe an error, the state of the connection become `CLOSE_WAIT`. I am not very sure if this will happen. This is the background.
So when the bad connection is used it throw a `Net` error `EOF`. But as I see the bad connection will not be really deleted as the code
```
func (p *stickyConnPool) Remove(cn *conn, _ error) error {
    defer p.mx.Unlock()
    p.mx.Lock()
    if p.closed {
        return errClosed
    }
    if p.cn == nil {
        panic("p.cn == nil")
    }
    if cn != nil && p.cn != cn {
        panic("p.cn != cn")
    }
    return nil
}
```
And when the `Multi` close the connection will be pushed back to the `connPool` by the `stickyConnPool` as a good one.
```
func (p *stickyConnPool) Close() error {
    defer p.mx.Unlock()
    p.mx.Lock()
    if p.closed {
        return errClosed
    }
    p.closed = true
    var err error
    if p.cn != nil {
        if p.reusable {
            err = p.put()
        } else {
            reason := errors.New("redis: sticky not reusable connection")
            err = p.remove(reason)
        }
    }
    return err
}
```
`p.reusable` is always true as the code
```
func (c *Client) Multi() *Multi {
    multi := &Multi{
        base: &baseClient{
            opt:      c.opt,
            connPool: newStickyConnPool(c.connPool, true),
        },
    }
    multi.commandable.process = multi.process
    return multi
}
```
Then what I said before will happen.
I am totally a new bird in github, and english is not my first language. So I am very happy for your response! Thank you!

**Comment by vmihailenco** [comment]
I think bad connection is removed by following code https://github.com/go-redis/redis/blob/master/multi.go#L48-L58. We also have a basic test https://github.com/go-redis/redis/blob/master/multi_test.go#L143-L168

But overall I agree that code is unnecessary complicated and subtle. I am thinking how to simplify it.

But if you could provide code to reproduce the issue I can try to dig deeper.

**Comment by dunixd** [comment]
This demo will display what I meet
```
package main

import (
    "gopkg.in/redis.v3"
    "time"
    "fmt"
)
var LOGGER_NAME = "REDIS_EXECUTOR"
func DEBUG(format string, args ...interface{}) {
    fmt.Printf("[DEBUG] [" + LOGGER_NAME + "] " + format + "\n", args...)
}
func INFO(format string, args ...interface{}) {
    fmt.Printf("[INFO ] [" + LOGGER_NAME + "] " + format + "\n", args...)
}
func WARN(format string, args ...interface{}) {
    fmt.Printf("[WARN ] [" + LOGGER_NAME + "] " + format + "\n", args...)
}
func ERROR(format string, args ...interface{}) {
    fmt.Printf("[ERROR] [" + LOGGER_NAME + "] " + format + "\n", args...)
}

func main() {
    INFO("Begin!")
    master := redis.NewClient(&redis.Options{
        Addr:        "127.0.0.1:28000",
        DB:          0,
        PoolTimeout: time.Millisecond * 1,
        IdleTimeout: time.Second * 120,
        PoolSize:    5,
    })
    pong, err := master.Ping().Result()
    INFO("master pong: %v, err: %v", pong, err)
    if err != nil {
        WARN("master.Ping err: %v", err)
        return
    }

    count := 0
    for {
        count++
        INFO("count: %d", count)
        func () error {
            tx, err := master.Watch("watch-key")
            if err != nil {
                WARN("Watch err: %v", err)
                return err
            }
            defer tx.Close()
            return nil
        }()
        time.Sleep(time.Second * 1)
    }

    INFO("End!")
}
```
First. Run it

> [INFO ] [REDIS_EXECUTOR] Begin!
> [INFO ] [REDIS_EXECUTOR] master pong: PONG, err: <nil>
> [INFO ] [REDIS_EXECUTOR] count: 1
> [INFO ] [REDIS_EXECUTOR] count: 2
> [INFO ] [REDIS_EXECUTOR] count: 3
> [INFO ] [REDIS_EXECUTOR] count: 4
> [INFO ] [REDIS_EXECUTOR] count: 5
> [INFO ] [REDIS_EXECUTOR] count: 6
> [INFO ] [REDIS_EXECUTOR] count: 7
> [INFO ] [REDIS_EXECUTOR] count: 8
> [INFO ] [REDIS_EXECUTOR] count: 9
> [INFO ] [REDIS_EXECUTOR] count: 10
> [INFO ] [REDIS_EXECUTOR] count: 11
> ...

Second. Restart redis server
Third. Then the program will display err forever

> 2016/03/01 19:05:29 redis: Unwatch failed: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [WARN ] [REDIS_EXECUTOR] Watch err: read tcp 127.0.0.1:60738->127.0.0.1:28000: read: connection reset by peer
> [INFO ] [REDIS_EXECUTOR] count: 44
> 2016/03/01 19:05:30 redis: Unwatch failed: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [WARN ] [REDIS_EXECUTOR] Watch err: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [INFO ] [REDIS_EXECUTOR] count: 45
> 2016/03/01 19:05:31 redis: Unwatch failed: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [WARN ] [REDIS_EXECUTOR] Watch err: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [INFO ] [REDIS_EXECUTOR] count: 46
> 2016/03/01 19:05:32 redis: Unwatch failed: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [WARN ] [REDIS_EXECUTOR] Watch err: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [INFO ] [REDIS_EXECUTOR] count: 47
> 2016/03/01 19:05:33 redis: Unwatch failed: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [WARN ] [REDIS_EXECUTOR] Watch err: write tcp 127.0.0.1:60738->127.0.0.1:28000: write: broken pipe
> [INFO ] [REDIS_EXECUTOR] count: 48

**Comment by dunixd** [comment]
When I call Client.Watch the process will call baseClient.process as this code
```
func (c *Multi) process(cmd Cmder) {
    if c.cmds == nil {
        c.base.process(cmd)
    } else {
        c.cmds = append(c.cmds, cmd)
    }
}
```
So when baseClient.process run to [https://github.com/go-redis/redis/blob/master/redis.go#L62-L62](https://github.com/go-redis/redis/blob/master/redis.go#L62-L62), baseClient.putConn will be called after, not Multi.putConn

**Comment by vmihailenco** [comment]
Thanks for the reproducer. I verified that it is fixed in https://github.com/go-redis/redis/pull/261 and added a test.

**Comment by vmihailenco** [comment]
@dunixd I just released v3.3.0 with fix. Please try it.

**Comment by dunixd** [comment]
OK, I am using it. Thank you very much!
